### PR TITLE
Dynamic reconfiguration of k3s on control planes

### PR DESCRIPTION
**Pull request by mraerino**
The idea is to allow for the config provisioner to re-run whenever the config changes and thereby apply it to the control plane node.

This is more of a PoC than anything and also a request for comments.

- How do you imagine dynamic reconfiguration of the control plane to work?
- Should there be a rollback mechanism in case the new config does not work?
- Can we do the nodes one-at-a-time to not cause control plane downtime if the config is faulty?
- Should we do a node restart instead of just restarting the k3s unit?
- How can we work around the fact that applying the config on node creation is different than when it is changed on an existing node?

_Related to https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/850_

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -642,7 +642,7 @@ variable "enable_wireguard" {
 }
 
 variable "control_planes_custom_config" {
-  type        = map(any)
+  type        = any
```
Comment (mysticaltech) [inline variables.tf:645]: Why this change?

File: control_planes.tf [hunk 2]
```diff
@@ -93,44 +129,34 @@ resource "null_resource" "control_planes" {
...
-      )
-    )
-
+    content     = yamlencode(local.k3s-config[each.key])
     destination = "/tmp/config.yaml"
   }
 
+  provisioner "remote-exec" {
+    inline = [
+      "mv /tmp/config.yaml /etc/rancher/k3s/config.yaml",
```
Comment (mysticaltech) [inline control_planes.tf:138]: Here we can use the same kind of custom logic used for k3s_registries to backup, and revert to the backed-up version if k3s is not successful in restarting.

**Comment by mysticaltech** [comment]
@mraerino Very interesting initial work, left a few comments, and I would like that we also tackle agent nodes in the same way, but let's corner control planes first indeed 🙏

As for your questions above, the methodology that was used for k3s_registries seems to work, so we create a backup either way if initial or not, and if k3s fails to start we restaure the backup. In cases of change (not initial), this will ensure we do not break a working cluster.

File: variables.tf (hunk 1 above)
Comment (mysticaltech) [reply variables.tf:645]: Yes, just saw, and merged it, makes sense! Thanks @mraerino 🙏

File: control_planes.tf (hunk 2 above)
Comment (mysticaltech) [reply control_planes.tf:138]: @mraerino Do you need help here? It's a good PR, just need this small tweak.

**Comment by Silvest89** [comment]
@mysticaltech This is taking way too long, I will be taking this over since I need it in my current cluster. And I have already been doing a lot of custom changes ;]

**Comment by mysticaltech** [comment]
@Silvest89  PR most welcome, please see the logic behind `k3s_registries` for inspiration. We basically need the same thing.

**Comment by mysticaltech** [comment]
@mraerino Thanks for starting this. @Silvest89 Built up the remaining needed functionality.

**Pull request closed without merging**